README.md
## PythonPackageSyncTool

PythonPackageSyncTool is a Python utility to "fix" requirements.txt.

It is used to make manipulation on requirements.txt that is produces by `pip freeze`

### Getting Help

### QuickStart
First of all it is better to install `alex_ber_utils` (this dependency is used in `setup.py`). 

```bash
pip3 install -U alex_ber_utils
```

##
Than:


```bash
pip3 install -U python-package-sync-tool
```

##
In some directory create ```config.yml```. Here is example 
(https://github.com/alex-ber/alpine-anaconda3/blob/master/python_package_sync_tool_config.yml)

```yaml
treeroot:
    source: installed.txt
    destination: rm.txt
    mutual_exclusion: True
    remove:
      - conda
      - conda-package-handling
      - icc-rt
      - intel-openmp
      - mkl
      - mkl-fft
      - mkl-random
      - mkl-service
      - pycosat
      - sip
      - tbb
    add:
  #    - graphviz==2.38.0
  #    - entrypoints==0.2.3
  #    - numpy==1.16.2
  #    - PyYAML==5.1
  #    - sasl==0.2.1
  #    - scipy==1.2.1
  #    - tables==3.4.4
  #    - terminado==0.8.2
  #    - tornado==5.1.1
  #    - Twisted==19.2.0
```

Create ``installed.txt`` with installed packages (in ```pip freeze``` format).

Now, you can run

```bash
python3 -m alexber.reqsync.data
```

You can also override parameters in the yaml file:

```bash
python3 -m alexber.reqsync.data --treeroot.add=some_new_package:1.0.0
```

shorthand

```bash
python3 -m alexber.reqsync.data --add=some_new_package:1.0.0
```

is also supported.


Of course, you can provide all parameters via CLI and not to use YAML file.

You can also supply ``--general.log`` (no shorthand here) parameter (you can do it also via ```config.yml```).
There is some default log configuration (to write everything to `stdout`).

`Note:` `config.yml` now also supports format that `alex-ber-utils` 
https://github.com/alex-ber/AlexBerUtils support for a long time.

Please, consult https://medium.com/analytics-vidhya/my-major-init-app-conf-module-1a5d9fb3998c for
full supported list of `general` parameters.

See documentation here

My `parser` module https://medium.com/analytics-vidhya/my-parser-module-429ed1457718

My `ymlparsers` module https://medium.com/analytics-vidhya/my-ymlparsers-module-88221edf16a6

My major `init_app_conf` module https://medium.com/analytics-vidhya/my-major-init-app-conf-module-1a5d9fb3998c


Note:

Semicolomn and not equal sign is used here due to Python limitaion of usage of equal sign in the value in the command line.

You can specified multiple packages using comma delimiter.

You can specify path to your config file using `--config_file` (or `--general.config_file`).
It can be absolute or relative.



##
Alternatively,

```bash
cd /opt/anaconda3/lib/python3.8/site-packages/alexber/reqsync/data/
```
Note: This is Path where you're actually install my utility, it can be different in your machine.

If you use venv it will look something like:

```bash
cd /opt/MyProject/venv/Lib/site-packages/alexber/reqsync
```
##


Alternatively you can create script file for yourself, named, say, driver.py:

```python3
#!/usr/bin/python3

import alexber.reqsync.app as app

if __name__ == "__main__":
   app.main()
```

Than create file config.yml near your script (see data/config.yml) or provide all parameter using command line
argruments. Use ':' in places where you should naturally write '==' (see explanation below).

Parameters 'source' and 'destination' are required. You should also provide (requirements) file for 'source'.

`mutual_exclusion` has default value True.



##
Now, type

```bash
python3 -m alexber.reqsync.data --add=some_new_package:1.0.0
```

or if you're using script (driver.py) go the directory with the script and type
```bash
./driver.py --add=some_new_package:1.0.0
```

or if you install my tool to Anaconda/Python/venv that has it's bin folder is in the Path
you can run  
```bash
python_package_sync_tool --add=some_new_package:1.0.0
```

or alternativley

you can run  
```bash
reqsync --add=some_new_package:1.0.0
```


This will run quick check whether package is not in remove list. If it is, the utility will fail. You can override this
beahivor by supplying `--mutual_exclusion=False`. 

Then, this will add some_new_package with version 1.0.0 to the requirements-dest.txt

Note:

Semicolomn and not equal sign is used here due to Python limitaion of usage of equal sign in the value in the command line.

You can specified multiple packages using comma delimiter.

You can specifiy path to your config file using `--config_file` (or `--general.config_file`).

It can be absolute or relative. If you're running using script (driver.py), that it can be relative to the directory 
whether you put your script. If you're running as the module (`python3 -m`), it can be relative to 
`/opt/anaconda3/lib/python3.8/site-packages/alexber/reqsync/data/` (exact path can be different, see above).  

##
You can supplied multiply packages by using comma:


```bash
python3 -m alexber.reqsync.data --add=some_new_package:1.0.0,another_new_package:2.0.0
```
or if you're using script (driver.py) go the directory with the script and type
```bash
./driver.py --add=some_new_package:1.0.0,another_new_package:2.0.0
```


### Installing from Github

```bash
python3 -m pip install -U https://github.com/alex-ber/PythonPackageSyncTool/archive/master.zip
```
Optionally installing tests requirements.

```bash
python3 -m pip install -U https://github.com/alex-ber/PythonPackageSyncTool/archive/master.zip#egg=python-package-sync-tool[tests]
```

Or explicitly:

```bash
wget https://github.com/alex-ber/PythonPackageSyncTool/archive/master.zip -O master.zip; unzip master.zip; rm master.zip
```
And then installing from source (see below).



### Installing from source
```bash
python3 -m pip install . # only installs "required"
```
```bash
python3 -m pip install .[test] # installs dependencies for tests
```

### Using Docker
`alexberkovich/python_package_sync_tool:latest`  contains all `python_package_sync_tool` dependencies.
This Dockerfile is very simple, you can take relevant part for you and put them into your Dockerfile.

##
Alternatively, you can use it as base Docker image for your project and add/upgrade 
another dependencies as you need.

For example:

```Dockerfile
FROM alexberkovich/python_package_sync_tool:latest

COPY requirements.txt etc/requirements.txt

RUN set -ex && \
    #latest pip,setuptools,wheel
    pip install --upgrade pip setuptools wheel && \
    pip install python_package_sync_tool 
    pip install -r etc/requirements.txt 

CMD ["/bin/sh"]
#CMD tail -f /dev/null
```

where `requirements.txt` is requirements for your project.

##

From the directory with setup.py
```bash
python3 setup.py test #run all tests
```
```bash
pytest
```



## Requirements


PythonPackageSyncTool requires the following modules.

* Python 3.7+

* alex-ber-utils==0.6.2

* alex-ber-utils[yml]==0.6.2

See https://github.com/alex-ber/AlexBerUtils

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

\#https://pypi.org/manage/project/python-package-sync-tool/releases/


## Unreleased
#### Changed
- Base docker image version to alexberkovich/alpine-anaconda3:0.2.1-slim.
alexberkovich/alpine-anaconda3:0.1.1 has some minor changes relative to alexberkovich/alpine-anaconda3:0.1.1.
See [https://github.com/alex-ber/alpine-anaconda3/blob/master/CHANGELOG.md] for details.


## [0.5.3] - 2020-12-17
### Changed
- `REAMDE.md`, requiremetes section.


## [0.5.2] - 2020-12-17
### Removed
- `alexber.reqsync.utils.app_conf` removed.

### Changed
- Pytest version updated (also some tests dependencies changed)
- alex-ber-utils dependency updated.
- Logger information was moved from `app.py` to `config.yml` file.
- `config.yml` now also supports format that `alex-ber-utils` 
https://github.com/alex-ber/AlexBerUtils support for a long time.


Please, consult https://medium.com/analytics-vidhya/my-major-init-app-conf-module-1a5d9fb3998c for
full supported list of `general` parameters.

See documentation here

My `parser` module https://medium.com/analytics-vidhya/my-parser-module-429ed1457718

My `ymlparsers` module https://medium.com/analytics-vidhya/my-ymlparsers-module-88221edf16a6

My major `init_app_conf` module https://medium.com/analytics-vidhya/my-major-init-app-conf-module-1a5d9fb3998c

- Logger is enabled, relative path is resolved correctly, and other little tweaks, see

Integrating Python’s logging and warnings packages 
[https://medium.com/analytics-vidhya/integrating-pythons-logging-and-warnings-packages-7ffd6f65e02d]

Making relative path to file to work 
[https://alex-ber.medium.com/making-relative-path-to-file-to-work-d5d0f1da67bf]
 
- `config.xml` - to remove package was added mkl*-related packagages and anaconda-specific packages.

- `REAMDE.md` was updated to reflect all changes above. 

### Added
- Optional Docker support added.  


## [0.5.1] - 2020-10-25
### Added
- Support for Python 3.8

### Changed
- 

## [0.4.3] - 2019-10-17
### Changed
- anaconda-navigator and conda-build added to config.yml.

## [0.4.2] - 2019-10-16
### Changed
- navigator-updater added to config.yml.

## [0.4.1] - 2019-05-30
### Changed
- Bug fix: adding packages before all existing one works incorrect.
- Removing alexber.reqsync.utils.parsers. It was fully duplicated by alexber.utils.parsers. 
So, all usage was change to the latest (part of alex-ber-utils).

##
## [0.3.1] - 2019-05-23
### Changed
- Bug fix: adding packages before all existing one works incorrect.

### Added
- Unit test for bug fix that check adding packages before all existing one.
- More detail assertion to integration tests.
- Unit-test for non-sorted requirements-src.
- Unit-test that check run with empty add and empty remvoe.
- Unit-test that check removing single package.
- Unit-test that check remove first package in requirements-src.
- Unit-test that check remove last package in requirements-src.
- Unit-test that check that empty lines in requirements-src are ignored.
- Unit-test that check correct usage of file input buffer and file output buffer.


## [0.2.11] - 2019-05-22
### Changed
- Dependency alex-ber-utils bumped up to 0.2.5.

## [0.2.8] - 2019-05-22
### Changed
- Dependency alex-ber-utils bumped up to 0.2.4.

## [0.2.6] - 2019-05-22
### Changed
- Dependency alex-ber-utils bumped up to 0.2.3.

## [0.2.5] - 2019-05-22
### Changed
- Fixed bug in setup.py, incorrect order between VERSION and UploadCommand (no tag was created on upload)
- Dependency alex-ber-utils bumped up to 0.2.2. 

## [0.2.4] - 2019-05-22
### Changed
- Adding dependency alex-ber-utils 0.2.1 to README.md.


## [0.2.3] - 2019-05-22
### Changed
- Upgrading urllib3, SQLAlchemy, pycrypto dependenies beacause of volnurabilities issues.

## [0.2.2] - 2019-05-22
### Changed
- Fixing python-package-sync-tool.
- Creating alias reqsync to python-package-sync-tool.
- Some minour fixed.


## [0.2.1] - 2019-05-22
### Changed
- Changing dependency version of alex-ber-utils to 0.2.1.


## [0.2.0] - 2019-05-22
### Changed
- Only bumping up version.

## [0.1.9] - 2019-05-22
### Changed
- Added alex-ber-utils as dependency. 
- Deleting old README-old.rst file.
- requirements-src.txt updated.
- Clarification added to README.md that alex_ber_utils should be installed first.
- formatting CHANGELOG.MD (minor fix)
- Added alternative of usage of python_package_sync_tool to README.md
- Fixing bug that --add is empty
- Factor out tests_data to seperate folder, use importlib.resources API.  
- Updated README-old.rst


## [0.1.8] - 2019-05-20
### Changed
- README.md change, key '-U' added to pip3 install.


## [0.1.7] - 2019-05-20
### Removed
- Some project cleanup.

### Changed
- CHANGELOG and REAMDE now use Markdown format.
- REAMDE totally rewritten.
- Fixing bugs in the core algorithm. Simplifying code.
- Fixing correct handling of package adding to the buttom of the list. 


## [0.1.6] - 2019-05-20
### Added
- `__init__.py` file added to alexber.reqsync.data.

## [0.1.5] - 2019-05-20
### Added
- Small tool to sync package from different machines.








tests/conftest.py
#one-time setup
import os
cwd = os.getcwd()
if cwd.endswith('PythonPackageSyncTool'):
    os.chdir(os.path.join(cwd, 'data'))

#see https://docs.pytest.org/en/latest/mark.html
#see https://docs.pytest.org/en/latest/example/simple.html#control-skipping-of-tests-according-to-command-line-option
def pytest_configure(config):
    config.addinivalue_line(
        "markers", "it:"
    )


tests/test_reqsync/app_test.py
import logging
logger = logging.getLogger(__name__)

import os
from pathlib import Path
import pytest

import alexber.reqsync.app as app
from alexber.reqsync import conf
from alexber.reqsync.app import init_app_conf, _PythonPackageSyncToolConfParser

from contextlib import ExitStack

_real_parse_config = init_app_conf.parse_config

from importlib.resources import path
from alexber.utils.parsers import is_empty

def _extract_pck(element):
    if element is None or is_empty(element):
        return ''
    elif '==' not in element:
        pck = element
    else:
        pck, _ = element.split('==')
    return pck


def validate_result(input_path, output_path, **kwargs):
    assert not is_empty(input_path)
    assert not is_empty(output_path)

    with open(output_path, 'rt') as f:
        out_requirements = [line.rstrip() for line in f.readlines()]

    sorted_requirements = sorted(out_requirements, key=lambda line: line.split('==')[0].casefold())
    pytest.assume(sorted_requirements==out_requirements)

    new_lines = kwargs.get('new_lines', [])
    for line in new_lines:
        pytest.assume(line in sorted_requirements)

    removed_lines = kwargs.get('removed_lines', [])
    for line in removed_lines:
        pytest.assume(line not in sorted_requirements)


    with open(input_path, 'rt') as f:
        in_requirements = [line.rstrip() for line in f.readlines()]
    for line in in_requirements:
        if is_empty(line):
            continue
        if line in new_lines:
            continue
        if _extract_pck(line) in removed_lines:
            continue
        pytest.assume(line in sorted_requirements)

#_log_config

@pytest.fixture
def initappFixture(mocker):
    app.logger = logging.getLogger(app.__name__)


@pytest.fixture
def initCwd(mocker):
    cwd = os.getcwd()
    #.../PythonPackageSyncTool/data
    new_cwd = Path(__file__).parent.parent.parent /'data'
    os.chdir(str(new_cwd))
    yield cwd
    os.chdir(cwd)


def test_main(request, mocker, initCwd):
    logger.info(f'{request._pyfuncitem.name}()')

    _parse_config_return_value = None

    def _mock_parse_config(argumentParser=None, args=None):
        ret = _real_parse_config(argumentParser, args)
        nonlocal _parse_config_return_value
        _parse_config_return_value = ret
        return ret

    mocker.patch.object(app, 'fixabscwd', autospec=True, spec_set=True)

    #mocker.spy(init_app_conf, 'parse_config')
    mocker.patch.object(init_app_conf, 'parse_config', side_effect=_mock_parse_config, autospec=True, spec_set=True)
    mocker.patch.object(app, 'run', autospec=True, spec_set=True)


    arg_source = 'source=requirements-src.txt'
    arg_destination = 'destination=requirements-dest.txt'
    arg_remove = 'remove=datashape,menuinst'

    exp_source = f'--treeroot.{arg_source}'
    exp_destination = f'--treeroot.{arg_destination}'
    exp_remove = f'--treeroot.{arg_remove}'

    argsv = f'--{arg_source} ' \
    f'--{arg_destination} ' \
    f'--{arg_remove} ' \
        .split()

    app.main(argsv)

    pytest.assume(init_app_conf.parse_config.call_count == 1)
    _, kwargs =  init_app_conf.parse_config.call_args
    source, destination, remove, _ = kwargs['args']
    pytest.assume( (exp_source, exp_destination, exp_remove) == (source, destination, remove) )

    pytest.assume(app.run.call_count == 1)
    _, run_d = app.run.call_args

    pytest.assume(_parse_config_return_value == run_d)


def test_main_unofficial_api(request, mocker, initCwd):
    logger.info(f'{request._pyfuncitem.name}()')

    _parse_config_return_value = None

    def _mock_parse_config(argumentParser=None, args=None):
        ret = _real_parse_config(argumentParser, args)
        nonlocal _parse_config_return_value
        _parse_config_return_value = ret
        return ret

    mocker.patch.object(app, 'fixabscwd', autospec=True, spec_set=True)

    #mocker.spy(init_app_conf, 'parse_config')
    mocker.patch.object(init_app_conf, 'parse_config', side_effect=_mock_parse_config, autospec=True, spec_set=True)
    mocker.patch.object(app, 'run', autospec=True, spec_set=True)


    arg_source = 'treeroot.source=requirements-src.txt'
    arg_destination = 'treeroot.destination=requirements-dest.txt'
    arg_remove = 'treeroot.remove=datashape,menuinst'

    exp_source = f'--{arg_source}'
    exp_destination = f'--{arg_destination}'
    exp_remove = f'--{arg_remove}'

    argsv = f'--{arg_source} ' \
    f'--{arg_destination} ' \
    f'--{arg_remove} ' \
    '--general.config.file=config.yml' \
        .split()

    app.main(argsv)

    pytest.assume(init_app_conf.parse_config.call_count == 1)
    _, kwargs =  init_app_conf.parse_config.call_args
    source, destination, remove, _, _ = kwargs['args']
    pytest.assume( (exp_source, exp_destination, exp_remove) == (source, destination, remove) )

    pytest.assume(app.run.call_count == 1)
    _, run_d = app.run.call_args

    pytest.assume(_parse_config_return_value == run_d)

def test_run(request, mocker, initappFixture):
    logger.info(f'{request._pyfuncitem.name}()')

    file_manager = ExitStack()

    pck = '.'.join(['tests_data', __package__, 'it'])

    exp_config_yml = file_manager.enter_context(
        path(pck, "config.yml"))
    exp_input = file_manager.enter_context(
        path(pck, 'requirements-src.txt'))
    exp_output = file_manager.enter_context(
        path(pck, 'requirements-dest.txt'))
    exp_removes = ['datashape','menuinst']


    d = {'config_file': str(exp_config_yml),
        'source': str(exp_input),
        'destination': str(exp_output),
        'remove':exp_removes,
        'add': None
             }

    app.run(**{conf.APP_KEY:d})

    validate_result(input_path=exp_input, output_path=exp_output, removed_lines=exp_removes)

def test_run_unsorted_src_req(request, mocker, initappFixture):
    logger.info(f'{request._pyfuncitem.name}()')

    file_manager = ExitStack()

    pck = '.'.join(['tests_data', __package__, 'unsorted_src_req'])

    exp_config_yml = file_manager.enter_context(
        path(pck, "config.yml"))
    exp_input = file_manager.enter_context(
        path(pck, 'requirements-src.txt'))
    exp_output = file_manager.enter_context(
        path(pck, 'requirements-dest.txt'))
    exp_removes = ['datashape','menuinst']


    d = {'config_file': str(exp_config_yml),
         'source': str(exp_input),
        'destination': str(exp_output),
        'remove':exp_removes,
        'add': None
             }

    with pytest.raises(ValueError, match='xpected to be sorted'):
        app.run(**{conf.APP_KEY:d})



@pytest.mark.parametrize(
     'source',

    [
        ('requirements-src.txt'),                   #no_change (basic)
        ('requirements-src-middle_empty_lines.txt'),
        ('requirements-src-first_empty_line.txt'),
        ('requirements-src-last_empty_lines.txt'),
        (None),                                     #_READ_BUFFER_SIZE / _WRITE_BUFFER_SIZE

    ]
)
def test_run_no_change(request, mocker, source, initappFixture):
    logger.info(f'{request._pyfuncitem.name}()')

    if source is None: #check correct usage of file input buffer and file output buffer.
        mocker.patch.object(app, '_READ_BUFFER_SIZE', new=40)
        mocker.patch.object(app, '_WRITE_BUFFER_SIZE', new=30)
        source = 'requirements-src.txt'

    file_manager = ExitStack()

    pck = '.'.join(['tests_data', __package__, 'it'])

    exp_config_yml = file_manager.enter_context(
        path(pck, "config.yml"))
    exp_input = file_manager.enter_context(
        path(pck, source))
    exp_output = file_manager.enter_context(
        path(pck, 'requirements-dest.txt'))


    d = {'config_file': str(exp_config_yml),
        'source': str(exp_input),
        'destination': str(exp_output),
        'remove':None,
        'add': None
             }

    app.run(**{conf.APP_KEY:d})

    validate_result(input_path=exp_input, output_path=exp_output)

@pytest.fixture
def init_default_parser_kwargs(request, mocker):
    request_param = {} if (
        not hasattr(request, 'param')) else request.param
    yield request_param


@pytest.fixture
def initparserFixture(mocker, init_default_parser_kwargs):
    default_parser_cls = init_app_conf.default_parser_cls
    default_parser_kwargs = init_app_conf.default_parser_kwargs

    init_app_conf.initConfig(default_parser_cls=_PythonPackageSyncToolConfParser, default_parser_kwargs=init_default_parser_kwargs)
    yield _PythonPackageSyncToolConfParser
    init_app_conf.initConfig(default_parser_cls=default_parser_cls, default_parser_kwargs=default_parser_kwargs)


@pytest.mark.parametrize(
     'exp_value, value, is_windows_path',

    [
        ('lxml==4.3.3', 'lxml:4.3.3', None),  #: is maksed as ==
        ('lxml==4.3.3', 'lxml==4.3.3', None), #== remains untouched
        (['lxml:4.3.3'], ['lxml:4.3.3'], None),  #will not happen in practice, so it's ok to be incorrect
                                           #it just shouldn't blow up
        (['lxml==4.3.3'], ['lxml==4.3.3'], None), #list
        ('C:\\', 'C:\\', True),    #Windows path, remain utntouched
        ('/etc/', '/etc/', False),   #Posix   path, remain utntouched
        ('config.yml', 'config.yml', None)

    ]
)

def test_mask_value(request, mocker, exp_value, value, is_windows_path,
                    init_default_parser_kwargs, initparserFixture):
    logger.info(f'{request._pyfuncitem.name}()')

    ret = init_app_conf.mask_value(value)

    if is_windows_path is None or not is_windows_path:
        pytest.assume(exp_value == ret)
    else:
        assert is_windows_path
        is_run_on_windows = os.name == 'nt'
        if is_run_on_windows:
            #exact comparasion
            pytest.assume(exp_value == ret)
        elif not is_run_on_windows:
            def is_path_exists(p):
                try:
                    pa=Path(p)
                    return pa.exists()
                except OSError:
                    return False

            #loose comparasion, it is enough that both paths doesn't exists
            is_exists = is_path_exists(ret)
            pytest.assume(not is_exists)
            is_exists = is_path_exists(value)
            pytest.assume(not is_exists)


@pytest.mark.parametrize('init_default_parser_kwargs', [{'implicit_convert':False},
                                                    ], indirect=True)

def test_mask_value_no_implicit_convert(request, mocker, init_default_parser_kwargs, initparserFixture):
    logger.info(f'{request._pyfuncitem.name}()')

    exp_value = 'lxml:4.3.3'
    ret = init_app_conf.mask_value(exp_value)
    pytest.assume(exp_value == ret)


if __name__ == "__main__":
    pytest.main([__file__])


alexber/reqsync/app.py
import logging.config

import io
import sys as _sys
from pathlib import Path
from argparse import ArgumentParser

#import alexber.reqsync.app_conf as conf
from alexber.utils.parsers import is_empty
import alexber.utils.ymlparsers as ymlparsers
from alexber.utils.mains import fixabscwd
import alexber.utils.init_app_conf as init_app_conf
from alexber.utils.init_app_conf import AppConfParser

from collections import deque

_READ_BUFFER_SIZE = 2 ** 16
_WRITE_BUFFER_SIZE = 2 ** 16

class conf(init_app_conf.conf):
    LOG_KEY = 'log'
    APP_KEY = 'treeroot'
    SOURCE_KEY = 'source'
    DEST_KEY= 'destination'
    MUTUAL_EXCLUSION_KEY = 'mutual_exclusion'
    RM_KEY = 'remove'
    ADD_KEY = 'add'


def _getSourceGen(filename, more_pck):
    #yield None  # gracefully handled adding package before all existing ones

    buffersize = _READ_BUFFER_SIZE
    with open(filename, 'rt') as f:
        while True:
            lines_buffer = f.readlines(buffersize)
            if not lines_buffer:
                break
            for line in lines_buffer:
                pck = line.rstrip()  # remove '\r'
                if pck:
                    yield pck

    if more_pck is not None:
        yield None  # to nullify prev_line

        for line in deque(more_pck):  # copy-constructor, more_pck is also changed outside, so
            # on first usage, we're creating the copy
            # (duplication is handled correctly by the code)
            pck = line.rstrip()  # remove '\r'
            if pck:
                yield pck






def _process_line(prev_line, cur_line, **kwargs):
    cur_pck, _ = ('', None) if cur_line is None else cur_line.split('==')
    prev_pck, _ = ('', None) if prev_line is None else prev_line.split('==')

    #lowercase them
    low_prev_line = None if prev_line is None else prev_line.casefold()
    low_cur_line = None if cur_line is None else cur_line.casefold()

    if (low_prev_line is not None) and (low_prev_line == low_cur_line) and not is_empty(low_cur_line):
        return None #duplicate line, ignore


    low_prev_pck = prev_pck.casefold()
    low_cur_pck = cur_pck.casefold()

    if (low_prev_line is not None) and (low_prev_pck > low_cur_pck) and not is_empty(low_cur_pck):
        raise ValueError("Source file expected to be sorted. Use sort utilities, for example.")


    if (is_empty(low_prev_pck)) and (low_prev_pck == low_cur_pck) and not is_empty(low_cur_pck):
        raise ValueError(f"Packages in the source file should be unique, but duplicate package {cur_pck} is found.")


    add_pckgs = kwargs.get(conf.ADD_KEY, None)
    rm_pckgs = kwargs.get(conf.RM_KEY, None)
    ret = deque([])
    is_append_curr_line = True

    #remove packages first
    while rm_pckgs is not None and not is_empty(rm_pckgs) and not is_empty(low_cur_pck):
        rm_pck = rm_pckgs[0]
        low_rm_pck = rm_pck.casefold()

        if low_cur_pck < low_rm_pck:
            is_append_curr_line = True
            break
        elif low_cur_pck == low_rm_pck:
            rm_pckgs.popleft()
            is_append_curr_line = False
            break
        else:
            rm_pckgs.popleft()
            continue


    # add packages
    while add_pckgs is not None and not is_empty(add_pckgs):
        add_line = add_pckgs[0]
        add_pck = _extract_pck(add_line)

        low_add_pck = add_pck.casefold()
        if is_empty(low_prev_pck) and (low_prev_pck < low_add_pck <= low_cur_pck): #adding new package at the head of file
            ret.append(add_line)
            is_append_curr_line = True
            add_pckgs.popleft()
        elif low_prev_pck < low_add_pck <= low_cur_pck:
            ret.append(add_line)
            is_append_curr_line = False
            add_pckgs.popleft()
        else:
            break

    if is_append_curr_line and not is_empty(cur_line):
        ret.append(cur_line)

    return ret

def _extract_pck(element):
    if element is None or is_empty(element):
        return ''
    elif '==' not in element:
        pck = element
    else:
        pck, _ = element.split('==')
    return pck


def _validate_mutual_exclusion(add_pckgs, rm_pckgs):
    if is_empty(add_pckgs) or is_empty(rm_pckgs):
        return

    s_add = {_extract_pck(element) for element in add_pckgs}
    for pck in rm_pckgs:
        if pck.casefold() in s_add:
            raise ValueError(f"Mutual_Exclusion enabled, but {pck} was found in both lists")



def _create_deque(pckgs):
    if pckgs is None:
        return None
    pckgs = [pck for pck in pckgs if pck is not None and pck]       #filter out None and ''
    ret_pckgs = None if pckgs is None else deque(sorted(pckgs, key=lambda s: s.casefold()))

    return ret_pckgs

def run(**kwargs):
    """
    This method recieved all conf params in kwargs.
    All unexpected values will be ignored.
    It is expected that value type is correct.
    No conversion on the value of the dict kwargs will be applied.
    This method will built playerA, playerB, engine,
    and run engine with these players.

    Please, consult alexber.rpsgame.app_conf in order to construct kwargs.
    Command-line argument and ini-file are suppored out of the box.
    JSON/YML, etc. can be easiliy handled also.
    """

    app_d = kwargs[conf.APP_KEY]

    logger.debug(f"app is {ymlparsers.as_str(app_d)}")

    src_f = app_d.get(conf.SOURCE_KEY, None)
    if src_f is None:
        raise ValueError(f'{conf.SOURCE_KEY} key should be defined')

    dest_f = app_d.get(conf.DEST_KEY, None)
    if dest_f is None:
        raise ValueError(f'{conf.DEST_KEY} key should be defined')

    add_pckgs = app_d.pop(conf.ADD_KEY, None)
    # Limitation: in-memory sorted
    add_pckgs = _create_deque(add_pckgs)

    rm_pckgs = app_d.pop(conf.RM_KEY, None)
    # Limitation: in-memory sorted
    rm_pckgs = _create_deque(rm_pckgs)

    app_d[conf.ADD_KEY] = add_pckgs
    app_d[conf.RM_KEY] = rm_pckgs

    is_mutual_exclusion = app_d.pop(conf.MUTUAL_EXCLUSION_KEY, True)
    if is_mutual_exclusion:
        _validate_mutual_exclusion(add_pckgs, rm_pckgs)

    full_src_path = Path(src_f).resolve()  # relative to cwd
    full_dest_path = Path(dest_f).resolve()  # relative to cwd
    sourceGen = _getSourceGen(full_src_path, add_pckgs)

    buffersize = _WRITE_BUFFER_SIZE
    lines_buffer = deque(maxlen=buffersize)
    prev_line = None

    with open(full_dest_path, 'wt') as f:
        for cur_line in sourceGen:
            lines = _process_line(prev_line, cur_line,
                                  **app_d)
            if lines is not None:
                lines_buffer.extend(lines)

            length = len(lines_buffer)
            if length ==  buffersize:
                # writelines ridiculously does not add newlines to the end of each line.
                f.writelines(map(lambda s: s + '\n', lines_buffer))
                lines_buffer.clear()
            prev_line = cur_line
        # writelines ridiculously does not add newlines to the end of each line.
        f.writelines(map(lambda s: s + '\n', lines_buffer))
        lines_buffer.clear()



def _log_config(**kwargs):
    log_d = kwargs.get(conf.GENERAL_KEY, {}).get(conf.LOG_KEY, {})
    if is_empty(log_d):
        log_d = _config["log_config"]

    logging.config.dictConfig(log_d)
    global logger
    logger = logging.getLogger(__name__)

    # logger.debug(f"{pprint.pprint(dd)}") #sort_dicts=False is available in Python 3.8
    logger.debug(f"Parsed log configurations is {ymlparsers.as_str(kwargs)}")

class _PythonPackageSyncToolConfParser(AppConfParser):
    def mask_value(self, value):
        def _is_path(w):
            p = Path(str(w))
            ret = p.drive!=''
            return ret

        def _fix(w):
            if _is_path(w):
                return w

            try:
                ret = w.replace(":", "==") \
                    if self.implicit_convert \
                    else w
                return ret
            except AttributeError:
                ret = w
            return ret

        ret = super().mask_value(value)
        ret = _fix(ret)
        return ret

    def _parse_white_list(self, default_d):
        white_list = [conf.GENERAL_KEY, conf.APP_KEY]
        return white_list

_ADAPT_KEYS =('source', 'destination', 'mutual_exclusion', 'remove', 'add')
_ADAPT_LIST_PREFIX = f'{conf.GENERAL_KEY}.{conf.LIST_ENSURE_KEY}'

def _adapt_arg_split(arg):
    if arg is None:
        return None, arg
    if not arg.startswith('--'):
        return None, arg

    ind = arg.find('=')

    if ind<0:
        key = arg
        value = ''
    else:
        key = arg[2:ind]
        value = arg[ind+1:]
    return key, value

def _adapt_arg(arg):
    key, value = _adapt_arg_split(arg)
    if key is None:
        return value

    ret = arg
    if key in _ADAPT_KEYS:
        ret = f'--{conf.APP_KEY}.{key}={value}'
    return ret

def _adapt_list_ensure(args):
    list_ensure_key = f'{conf.GENERAL_KEY}.{conf.LIST_ENSURE_KEY}'
    add_key = f'{conf.APP_KEY}.{conf.ADD_KEY}'
    rm_key =  f'{conf.APP_KEY}.{conf.RM_KEY}'
    buf = io.StringIO()
    buf.write(f'--{list_ensure_key}=')
    is_used = False

    for arg in args:
        key, value = _adapt_arg_split(arg)
        if list_ensure_key==key:
            return args
        if add_key==key :
            buf.write(f'{conf.APP_KEY}.{conf.ADD_KEY},')
            is_used = True
        elif rm_key==key :
            buf.write(f'{conf.APP_KEY}.{conf.RM_KEY},')
            is_used = True

    if is_used:
        args.append(buf.getvalue()[:-1])    #truncate last comma
    return args


def _adapt_conf(args=None):
    """
    This function is adapter layer for backproring of init_app_conf.

    :return:
    """

    # ymlparsers.initConfig()
    # init_app_conf.initConfig()

    argumentParser = ArgumentParser()
    #allow to use config.file
    argumentParser.add_argument("--config_file", nargs='?', dest='config_file', default='config.yml',
                                const='config.yml')

    init_app_conf.initConfig(default_parser_cls=_PythonPackageSyncToolConfParser)

    if args is None:
        # args default to the system args
        args = _sys.argv[1:]
    else:
        # make sure that args are mutable
        args = list(args)

    # allow to not to specify treesort
    args = [_adapt_arg(arg) for arg in args]
    # allow to not to specify general.listEnsure
    args = _adapt_list_ensure(args)

    dd = init_app_conf.parse_config(argumentParser=argumentParser, args=args)


    return dd




def main(args=None):
    """
    main method
    :param args: if not None, suppresses sys.args
    """
    logging.basicConfig(format='%(asctime)-15s %(levelname)s [%(name)s.%(funcName)s] %(message)s',
                        level=logging.INFO)
    logging.captureWarnings(True)

    dd = _adapt_conf(args)

    _log_config(**dd)
    del dd[conf.GENERAL_KEY]

    run(**dd)

#see https://terryoy.github.io/2016/05/short-ref-python-logging.html
_config = {
        "log_config": {
            "version": 1,
            "formatters": {
                "brief": {
                    "format": "%(message)s",
                },
                "detail": {
                    "format": "%(asctime)-15s %(levelname)s [%(name)s.%(funcName)s] %(message)s",
                    "datefmt": '%Y-%m-%d %H:%M:%S',
                },
            },
            "handlers": {
                "console": {
                    "class": "logging.StreamHandler",
                    "level": "DEBUG",
                    "formatter": "brief",
                },
                # "file": {
                #     "class": "logging.handlers.RotatingFileHandler",
                #     "filename": "dev.log",
                #     "level": "DEBUG",
                #     "formatter": "detail",
                # },
            },
            "root": {
                # "handlers": ["console", "file"],
                "handlers": ["console"],
                "level": "DEBUG",
            },
            "loggers": {
                "requests": {
                    # "handlers": ["file"],
                    "handlers": ["console"],
                    "level": "DEBUG",
                    "propagate": False,
                }
            },
        },
    }


if __name__ == "__main__":
    main()
